packages/ai/src/workflows/controllers.ts

```diff
@@ -1,105 +1,81 @@
-import { Artifact, CodeArtifact, PromptArtifact } from "../Artifact.js";
+import { PromptArtifact } from "../Artifact.js";
 import { IOShape } from "../LLMStepTemplate.js";
 import { adjustToFeedbackStep } from "../steps/adjustToFeedbackStep.js";
 import { fixCodeUntilItRunsStep } from "../steps/fixCodeUntilItRunsStep.js";
+import { generateCodeStep } from "../steps/generateCodeStep.js";
 import { matchStyleGuideStep } from "../steps/matchStyleGuideStep.js";
 import { Workflow } from "./Workflow.js";
 
-function handleCodeStep(
-  workflow: Workflow<any>,
-  prompt: PromptArtifact,
-  code: CodeArtifact | undefined,
-  currentStepType: typeof adjustToFeedbackStep | typeof matchStyleGuideStep,
-  nextStepType: typeof matchStyleGuideStep | undefined
-) {
-  // This means the step is done.
-  if (!code) {
-    if (nextStepType) {
-      const recentCode = workflow.getRecentValidCode();
-      if (!recentCode) {
-        throw new Error(
-          "Impossible state. Some code must have existed before."
-        );
-      }
-      workflow.addStep(nextStepType, { prompt, code: recentCode });
+function retryMinorFailures(workflow: Workflow<any>) {
+  workflow.addEventListener("stepFinished", ({ data: { step } }) => {
+    const state = step.getState();
+    if (state.kind !== "FAILED" || state.errorType !== "MINOR") {
+      return;
     }
-    // else, is done with run
-    return;
-  }
-
-  if (code.value.type !== "success") {
-    workflow.addStep(fixCodeUntilItRunsStep, { code });
-  } else {
-    workflow.addStep(currentStepType, { prompt, code });
-  }
-}
-
-function handleSuccessfulCode(
-  workflow: Workflow<any>,
-  prompt: PromptArtifact,
-  code: CodeArtifact,
-  nextStep: typeof adjustToFeedbackStep
-) {
-  if (code.value.type !== "success") {
-    workflow.addStep(fixCodeUntilItRunsStep, { code });
-  } else {
-    workflow.addStep(nextStep, { prompt, code });
-  }
+    workflow.addRetryOfPreviousStep();
+  });
 }
 
 export function fixAdjustRetryLoop<Shape extends IOShape>(
   workflow: Workflow<Shape>,
   prompt: PromptArtifact
 ) {
-  workflow.addEventListener("stepFinished", ({ data: { step } }) => {
-    const codeArtifact = step.getOutputs()["code"] as Artifact | undefined;
-    if (codeArtifact && codeArtifact.kind !== "code") {
-      throw new Error("Impossible state");
-    }
-    const code = codeArtifact?.kind === "code" ? codeArtifact : undefined;
+  // GENERATE
```

Comment:
The code below is highly repetitive (the diff is not very readable, btw, I recommend opening the full file), but I think even in this form it's still a win over the previous version. Repetitive simple code is better than dense complicated code.

The main reason why it's repetitive is because `LLMStepInstance` and `LLMStepTemplate` (or just one of them? not sure) are both covariant and contravariant... So when I try to run `addRule` in a loop with slightly different `after: template`, types don't pass.

I hope that I'll be able to clean this up, but it's not trivial. Probably will require going from `LLMStepInstance` to something like `LLMStepData`; I mean, I'm pretty sure that `execute` callback is the cause of contravariance which is the cause of type inflexibility. So if we pass around some simpler structure which doesn't point to the `execute` callback then it should fix the problem.